documentation/managing_modules.md

```diff
@@ -0,0 +1,183 @@
+# 🧪 Managing modules in Bolt projects
+
+Bolt version 2.30.0 introduced module workflows for Bolt projects that improve
+dependency management and make it easier to add new modules to your Bolt
+project. These workflows, which include the `bolt module install`, `bolt module
+add`, and `bolt project migrate` commands, are experimental.
+
+## What's changed?
+
+- **Module installation directory:** Bolt now installs modules into the
+  `.modules/` directory instead of `modules/`. Avoid committing `.modules/` to
+  source control. Your users can download your Bolt project and use the `bolt
+  module install` command to download the required modules from the forge.
+- **Module configuration:** Your `bolt-project.yaml` file contains a `modules`
+  key that lists the direct module dependencies of your Bolt project. Unless you
+  need to pin a module to a specific version, avoid editing the `modules` list
+  directly. Instead, use the `bolt module add` command.
+- **Puppetfile:** The Puppet file is now a lock file. Bolt generates a
+  Puppetfile each time you modify your modules with a Bolt command. Do not edit
+  the Puppetfile directly. Instead, use Bolt commands to manage your modules,
+  and rely on Bolt to manage the Puppetfile. You can compare a Puppetfile to a
+  `Gemfile.lock` file in Ruby, or a `yarn.lock` file in Yarn.
+- **Module path:** The new module path is `['modules']` and Bolt always appends
+  `.modules` to the module path.
+
+## How it works
+
+When you install a Forge module, Bolt 
+- installs the latest version of the module along with any dependencies required
+  by the module.
+- adds entries for the module to the `modules` key in your `bolt-project.yaml`,
+  and installs the module and its dependencies into a `.modules` directory in
+  your project. 
+- generates a Puppetfile, which lists all of a project's required modules and
+  their versions.
+
+For example, if you create a project named `example_project` and install the
+`apache` module, the `modules` key in your `bolt-project.yaml` file looks
+something like this:
+
+```yaml
+name: example_project
+modules:
+- name: puppetlabs-apache
+```
+
+The project's Puppetfile lists the `apache` module and all of its
+dependencies:
+
+```puppet
+# This Puppetfile is managed by Bolt. Do not edit.
+mod "puppetlabs-apache", "5.5.0"
+mod "puppetlabs-stdlib", "6.4.0"
+mod "puppetlabs-concat", "6.2.0"
+mod "puppetlabs-translate", "2.2.0"
+```
+
+And the `.modules` directory contains all of the installed modules:
+
+```shell
+$ ls modules
+apache    concat    stdlib    translate
+```
+
+## Opting in
+
+Use `bolt project migrate` to migrate a Bolt project to the new workflows. 
```

Comment:
`bolt project migrate` won't always work, or a user may not want to use the command for their own reasons. We should have a section describing how to manually opt in. We should recommend the user be using a VCS for their project. Assuming you have a Bolt project with a `bolt-project.yaml`, these are the steps they'll need to take:

- Add a `modules` key to the `bolt-project.yaml`
- Add any modules you want to manage with Bolt from your Puppetfile to the `modules` key. Only list modules that you use directly and not their dependencies. Only Puppet Forge modules are supported.
- Delete the Puppetfile.
- Delete the modules you are managing, plus their dependencies, from the modulepath (this is likely `modules` for most users).
- Move all remaining modules (likely in `site-modules` / `site`) to the `modules` directory.
- If you have configured `modulepath` in `bolt-project.yaml`, unconfigure it.
- Run `bolt module install` to install the managed modules.

Reply:
I was trying to avoid describing this workflow, because it seems like a less happy path than just using the `migrate` and `add` commands, and it makes it more difficult to be really clear about what a user needs to do to get going. 

I think we often fall into the trap of trying to document every possible way of doing something. Could you give me an example of a case where the migrate command won't work? Obviously, we want to document it if it's going to be an issue for a large group of users, but maybe it's a case of covering that use case in a separate doc?

Reply:
I think the most likely scenario is if an existing Puppetfile cannot be parsed, preventing a user from using `project migrate`. Per the chat in channel, I think we can hold off on describing this until/if users keep asking for it.